File: app/routers/recurring_items.py
from datetime import date

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import func
from sqlalchemy.orm import Session, joinedload

from app.database import get_db
from app.models import Category, LedgerTransaction, RecurringItem
from app.schemas.card import CardBrief
from app.schemas.ledger import (
    CategoryBrief,
    PaymentMethodBrief,
    RecurringGenerateItem,
    RecurringGenerateRequest,
    RecurringGenerateResponse,
    RecurringItemCreate,
    RecurringItemResponse,
    RecurringItemUpdate,
)

router = APIRouter(prefix="/recurring-items", tags=["recurring-items"])


def _recurring_item_query(db: Session):
    return db.query(RecurringItem).options(
        joinedload(RecurringItem.category).joinedload(Category.parent),
        joinedload(RecurringItem.payment_method),
        joinedload(RecurringItem.account),
        joinedload(RecurringItem.to_account),
        joinedload(RecurringItem.card),
    )


def _serialize_recurring_item(item: RecurringItem) -> RecurringItemResponse:
    parent_name = item.category.parent.name if item.category and item.category.parent else None
    card_brief = None
    if item.card:
        card_brief = CardBrief(
            id=item.card.id,
            name=item.card.name,
            card_type=item.card.card_type,
            institution=item.card.institution,
            last_four=item.card.last_four,
        )
    return RecurringItemResponse(
        id=item.id,
        type=item.type,
        amount=item.amount,
        category=CategoryBrief(
            id=item.category.id,
            name=item.category.name,
            parent_name=parent_name,
        ),
        payment_method=PaymentMethodBrief(id=item.payment_method.id, name=item.payment_method.name)
        if item.payment_method
        else None,
        account_id=item.account_id,
        to_account_id=item.to_account_id,
        account_name=item.account.name if item.account else None,
        to_account_name=item.to_account.name if item.to_account else None,
        card=card_brief,
        merchant=item.merchant,
        memo=item.memo,
        frequency=item.frequency,
        day_of_month=item.day_of_month,
        is_active=item.is_active,
    )


@router.get("", response_model=list[RecurringItemResponse])
def list_recurring_items(db: Session = Depends(get_db)):
    items = _recurring_item_query(db).order_by(RecurringItem.id).all()
    return [_serialize_recurring_item(item) for item in items]


@router.post("", response_model=RecurringItemResponse, status_code=201)
def create_recurring_item(payload: RecurringItemCreate, db: Session = Depends(get_db)):
    item = RecurringItem(**payload.model_dump())
    db.add(item)
    db.commit()
    item = _recurring_item_query(db).filter(RecurringItem.id == item.id).one()
    return _serialize_recurring_item(item)


@router.patch("/{item_id}", response_model=RecurringItemResponse)
def update_recurring_item(item_id: int, payload: RecurringItemUpdate, db: Session = Depends(get_db)):
    item = db.get(RecurringItem, item_id)
    if not item:
        raise HTTPException(status_code=404, detail="정기 항목을 찾을 수 없습니다.")
    for key, value in payload.model_dump(exclude_unset=True).items():
        setattr(item, key, value)
    db.commit()
    item = _recurring_item_query(db).filter(RecurringItem.id == item.id).one()
    return _serialize_recurring_item(item)


@router.delete("/{item_id}", status_code=204)
def delete_recurring_item(item_id: int, db: Session = Depends(get_db)):
    item = db.get(RecurringItem, item_id)
    if not item:
        raise HTTPException(status_code=404, detail="정기 항목을 찾을 수 없습니다.")
    db.delete(item)
    db.commit()


@router.post("/generate", response_model=RecurringGenerateResponse)
def generate_recurring_items(payload: RecurringGenerateRequest, db: Session = Depends(get_db)):
    items = _recurring_item_query(db).filter(RecurringItem.is_active.is_(True)).all()
    generated = 0
    skipped = 0
    results: list[RecurringGenerateItem] = []

    for item in items:
        exists = (
            db.query(LedgerTransaction)
            .filter(
                LedgerTransaction.type == item.type,
                LedgerTransaction.category_id == item.category_id,
                LedgerTransaction.amount == item.amount,
                func.strftime("%Y", LedgerTransaction.transaction_date) == str(payload.year),
                func.strftime("%m", LedgerTransaction.transaction_date) == f"{payload.month:02d}",
                LedgerTransaction.memo == (item.memo or f"정기 {item.type}"),
            )
            .first()
        )
        if exists:
            skipped += 1
            results.append(RecurringGenerateItem(recurring_item_id=item.id, status="already_exists"))
            continue

        day = min(item.day_of_month, 28)
        tx_date = date(payload.year, payload.month, day)
        tx = LedgerTransaction(
            transaction_date=tx_date,
            type=item.type,
            amount=item.amount,
            category_id=item.category_id,
            payment_method_id=item.payment_method_id,
            account_id=item.account_id,
            to_account_id=item.to_account_id,
            card_id=item.card_id,
            merchant=item.merchant,
            memo=item.memo or f"정기 {item.type}",
            is_fixed=True,
        )
        db.add(tx)
        db.flush()
        generated += 1
        results.append(
            RecurringGenerateItem(
                recurring_item_id=item.id,
                ledger_transaction_id=tx.id,
                status="created",
            )
        )

    db.commit()
    return RecurringGenerateResponse(generated=generated, skipped=skipped, items=results)

File: src/lib/services/recurring-items.ts
import Decimal from "decimal.js";

import { prisma } from "@/lib/db";
import { toDecimal } from "@/lib/decimal";
import {
  applySqliteDateRange,
  monthSqliteDateBounds,
} from "@/lib/sqlite-date-filter";
import { getQuote, RateLimitError } from "@/lib/external/market-data";
import { ASSET_CLASS_DEPOSIT } from "@/lib/utils";

export async function refreshPrices(holdingIds?: number[]) {
  const holdings = await prisma.holding.findMany({
    where: {
      quantity: { gt: 0 },
      asset_class: { not: ASSET_CLASS_DEPOSIT },
      ...(holdingIds?.length ? { id: { in: holdingIds } } : {}),
    },
  });

  let updated = 0;
  const failed: Array<{ holding_id: number; symbol: string; reason: string }> = [];

  for (const holding of holdings) {
    try {
      const quote = await getQuote(holding.symbol);
      await prisma.holding.update({
        where: { id: holding.id },
        data: {
          last_market_price: quote.price,
          last_price_updated_at: new Date(),
          ...(quote.name ? { name: quote.name } : {}),
        },
      });
      updated += 1;
    } catch (error) {
      if (error instanceof RateLimitError) {
        failed.push({
          holding_id: holding.id,
          symbol: holding.symbol,
          reason: "시세 서버 요청 한도 초과",
        });
        break;
      }
      failed.push({
        holding_id: holding.id,
        symbol: holding.symbol,
        reason: "시세 조회 실패",
      });
    }
  }

  return { updated, failed };
}

export async function generateRecurringItems(year: number, month: number) {
  const items = await prisma.recurringItem.findMany({
    where: { is_active: true },
    include: {
      category: { include: { parent: true } },
      payment_method: true,
      account: { select: { name: true } },
      to_account: { select: { name: true } },
      card: true,
    },
    orderBy: { id: "asc" },
  });

  let generated = 0;
  let skipped = 0;
  const results: Array<{
    recurring_item_id: number;
    ledger_transaction_id?: number;
    status: string;
  }> = [];

  const { from, to } = monthSqliteDateBounds(year, month);

  for (const item of items) {
    const memo = item.memo ?? `정기 ${item.type}`;
    const existing = await prisma.ledgerTransaction.findFirst({
      where: await applySqliteDateRange(
        {
          type: item.type,
          category_id: item.category_id,
          amount: item.amount,
          memo,
        },
        "ledger_transactions",
        from,
        to,
      ),
    });

    if (existing) {
      skipped += 1;
      results.push({ recurring_item_id: item.id, status: "already_exists" });
      continue;
    }

    const day = Math.min(item.day_of_month, 28);
    const txDate = new Date(year, month - 1, day);

    const tx = await prisma.ledgerTransaction.create({
      data: {
        transaction_date: txDate,
        type: item.type,
        amount: item.amount,
        category_id: item.category_id,
        payment_method_id: item.payment_method_id,
        account_id: item.account_id,
        to_account_id: item.to_account_id,
        card_id: item.card_id,
        merchant: item.merchant,
        memo,
        is_fixed: true,
      },
    });

    generated += 1;
    results.push({
      recurring_item_id: item.id,
      ledger_transaction_id: tx.id,
      status: "created",
    });
  }

  return { generated, skipped, items: results };
}
